Expo1900/Expo1900/Model/Decoder.swift

```diff
@@ -0,0 +1,26 @@
+//
+//  JsonDecoder.swift
+//  Expo1900
+//
+//  Created by yyss99, kyungmin on 2023/07/03.
+//
+
+import UIKit
+
+struct Decoder {
+    static func decodeJson<Data: Decodable>(from dataAssetName: String) throws -> Data {
```

Comment:
왜 `Data: Decodable`으로 표현하셨나요? `Decodable` 프로토콜 자체를 제네릭으로 넣으면 안되는걸까요? 만약 아래와 같이 바꾸면 어떤 일이 벌어지나요?
```suggestion
    static func decodeJson(from dataAssetName: String) throws -> Decodable {
```

Reply:
`Decodable`프로토콜 자체를 제네릭으로 넣으면 `Decodable`네임 자체가 제네릭 타입으로 돼서 `Decodable`프로토콜을 채택한 타입이라고 명시해줄 수 없는 것 같습니다😭
저희가 제네릭에 대해 잘못 이해하고 있는 걸까요?

Reply:
앗 수정했어요 코멘트 다시 확인 부탁드릴게요~

Reply:
```swift
let decodeData = try jsonDecoder.decode(T.self, from: dataAsset.data)
```
말씀해주신대로 바꾸면 위의 코드에서 DTO객체 타입을 찾을 수 가 없습니다😭

Reply:
프로토콜을 반환 타입으로 사용하면 확장성이 좋은 코드를 작성 할 수 있고,  의존성 주입을 통해 코드의 결합도를 낮출 수 있다는 것에 대해 알게됐습니다. 하지만 `decode`메서드의 `parameter`가 메타 타입 이기때문에 `Decodable.self`를 하면 프로토콜이 나오기 때문에 저희 코드에서는 적용할 수 없었습니다😭
```swift
enum Decoder {
    static func decodeJson(from dataAssetName: String) throws -> Decodable {
        let jsonDecoder: JSONDecoder = JSONDecoder()

        guard let dataAsset = NSDataAsset(name: dataAssetName) else {
            throw DecodeError.searchNoFile
        }

        do {
            let decodeData = try jsonDecoder.decode(Decodable.self, from: dataAsset.data)

            return decodeData
        } catch {
            throw DecodeError.jsonDecodeError
        }
    }
}
```